salt/modules/debbuild.py

```diff
@@ -416,20 +442,66 @@ def make_repo(repodir, keyid=None, env=None):
         fow.write('{0}'.format(repocfg_dists))
 
+    local_fingerprint = None
+    local_keyid = None
+
     if keyid is not None:
         with salt.utils.fopen(repoconfdist, 'a') as fow:
             fow.write('SignWith: {0}\n'.format(keyid))
 
+        # gpg keys should have been loaded as part of setup
+        # retrieve specified key, obtain fingerprint and preset passphrase
+        local_keys = __salt__['gpg.list_keys'](user=runas, gnupghome=gnupghome)
+        for gpg_key in local_keys:
+            if keyid == gpg_key['keyid'][8:]:
+                local_fingerprint = gpg_key['fingerprint']
+                local_keyid = gpg_key['keyid']
+                break
+
+        if local_keyid is None:
+            raise SaltInvocationError(
+                '\'keyid\' was not found in gpg keyring'
+            )
+
+        # preset passphase and interaction with gpg-agent
+        gpg_info_file = '{0}/gpg-agent-info-salt'.format(gnupghome)
+        with salt.utils.fopen(gpg_info_file, 'r') as fow:
+            gpg_raw_info = fow.readlines()
+
+        for gpg_info_line in gpg_raw_info:
+            gpg_info = gpg_info_line.split('=')
+            gpg_info_dict = {gpg_info[0]: gpg_info[1]}
+            __salt__['environ.setenv'](gpg_info_dict, update_minion=True)
+            break
+
+        if use_passphrase:
+            cmd = '/usr/lib/gnupg2/gpg-preset-passphrase --verbose --forget {0}'.format(local_fingerprint)
+            __salt__['cmd.run'](cmd, runas=runas)
+
+            phrase = __salt__['pillar.get']('gpg_passphrase')
+            cmd = '/usr/lib/gnupg2/gpg-preset-passphrase --verbose --preset --passphrase "{0}" {1}'.format(phrase, local_fingerprint)
+            __salt__['cmd.run'](cmd, runas=runas)
+
     repocfg_opts = _get_repo_options_env(env)
     repoconfopts = os.path.join(repoconf, 'options')
     with salt.utils.fopen(repoconfopts, 'w') as fow:
         fow.write('{0}'.format(repocfg_opts))
 
     for debfile in os.listdir(repodir):
+        abs_file = os.path.join(repodir, debfile)
         if debfile.endswith('.changes'):
-            cmd = 'reprepro -Vb . include {0} {1}'.format(codename, os.path.join(repodir, debfile))
-            __salt__['cmd.run'](cmd, cwd=repodir)
+            os.remove(abs_file)
+
+        if debfile.endswith('.dsc'):
+            cmd = 'debsign --re-sign -k {0} {1}'.format(keyid, os.path.join(repodir, abs_file))
+            __salt__['cmd.run'](cmd, cwd=repodir, use_vt=True)
+            cmd = 'reprepro --ignore=wrongdistribution --component=main -Vb . includedsc {0} {1}'.format(codename, abs_file)
+            __salt__['cmd.run'](cmd, cwd=repodir, use_vt=True)
 
         if debfile.endswith('.deb'):
-            cmd = 'reprepro -Vb . includedeb {0} {1}'.format(codename, os.path.join(repodir, debfile))
-            __salt__['cmd.run'](cmd, cwd=repodir)
+            cmd = 'reprepro --ignore=wrongdistribution --component=main -Vb . includedeb {0} {1}'.format(codename, abs_file)
```

Comment:
Should we be evaluating the existence of these shell commands before we run them?